### PR TITLE
Option -gui-address should accept scheme prefixes

### PR DESCRIPTION
**Pull request by calmh**
This is ugly and I'm tired

### CODE REVIEW

**Comment by calmh** [comment]
Actually, it's probably cleaner to have config handle this internally, similarly to the RawPath/Path() deal. Hang on a moment.

**Comment by calmh** [comment]
I think that's cleaner.

File: cmd/syncthing/main.go [hunk 1]
```diff
@@ -910,48 +913,18 @@ func setupGUI(mainSvc *suture.Supervisor, cfg *config.Wrapper, m *model.Model, a
...
-		if cfg.Options().StartBrowser && !noBrowser && !stRestarting {
-			urlOpen := fmt.Sprintf("%s://%s/", proto, net.JoinHostPort(hostOpen, strconv.Itoa(addr.Port)))
-			// Can potentially block if the utility we are invoking doesn't
-			// fork, and just execs, hence keep it in it's own routine.
-			go openURL(urlOpen)
-		}
+	if cfg.Options().StartBrowser && !noBrowser && !stRestarting {
+		// Can potentially block if the utility we are invoking doesn't
+		// fork, and just execs, hence keep it in it's own routine.
+		go openURL(guiCfg.URL())
```
Comment (AudriusButkevicius) [inline cmd/syncthing/main.go]: If someone specified `:8080` this will become `0.0.0.0:8080` which is invalid.

Comment (calmh) [inline cmd/syncthing/main.go]: Actually, that works just fine for me which why I removed some of the 0.0.0.0 → 127.0.0.1 conversions we already had. That, is, my browser accepts http://0.0.0.0:8080/ and connects to localhost for it. I haven't tested it extensively in other browsers or operating systems though... I'll do a quick check that it works at least in Windows/IE, and if it doesn't I'll reinstate the conversion.

Comment (AudriusButkevicius) [inline cmd/syncthing/main.go]: Tells me ERR_INVALID_ADDRESS on Chrome on Windows.

Comment (calmh) [inline cmd/syncthing/main.go]: All right, then I'll do what it does above, and as a last step convert 0.0.0.0 → 127.0.0.1 and [::] to [::1].

**Comment by calmh** [comment]
There, I think.

File: cmd/syncthing/gui.go [hunk 2]
```diff
@@ -125,11 +120,13 @@ func (s *apiSvc) getListener(cfg config.GUIConfiguration) (net.Listener, error)
 		},
 	}
 
-	rawListener, err := net.Listen("tcp", cfg.Address)
+	rawListener, err := net.Listen("tcp", guiCfg.Address())
 	if err != nil {
 		return nil, err
 	}
 
+	l.Infoln("Starting web GUI on", guiCfg.URL())
```
Comment (AudriusButkevicius) [inline cmd/syncthing/gui.go]: Now this is wrong, as it hides the fact that the GUI is out in the open.

File: lib/config/config.go [hunk 3]
```diff
@@ -288,12 +289,72 @@ func (orig OptionsConfiguration) Copy() OptionsConfiguration {
...
+		Scheme: "http",
+		Host:   c.Address(),
+		Path:   "/",
+	}
+
+	if c.UseTLS() {
+		u.Scheme = "https"
+	}
+
+	if strings.HasPrefix(u.Host, ":") {
```
Comment (AudriusButkevicius) [inline lib/config/config.go]: We should be good citizens, and use net.SplitHost, net.ParseIP, ip.IsUnspecified, as this whole string business is fragile. You already have c.Address() so thats half of the job done.

File: cmd/syncthing/gui.go (hunk 2 above)
Comment (calmh) [inline cmd/syncthing/gui.go]: That was the case before also. There's another "API listening on ...", and I'm not sure we really need both, but this one does print the URL and the other prints the listen address...

File: lib/config/config.go (hunk 3 above)
Comment (calmh) [inline lib/config/config.go]: That's a "meh" to me as this is mostly for printing purposes and does the right thing for the formats we expect (there's even a test). If we do that we need to return an error, and that bubbles around everywhere. I'd rather just let it fail when used, as we'll some error when trying to listen on "234.#€./%.47:123" before even getting to the point where we print the URL. :)

File: cmd/syncthing/gui.go (hunk 2 above)
Comment (AudriusButkevicius) [inline cmd/syncthing/gui.go]: Lets get rid of this then.

Comment (calmh) [inline cmd/syncthing/gui.go]: How would you like to print it? Both are useful - one gives me a double-clickable link in the terminal, the other tells the truth. ;)

**Comment by calmh** [comment]
How about that? ^

**Pull request merged**